src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/TrustHelper.cs

```diff
@@ -947,30 +947,61 @@ private static void ValidateTrustAttribute(TRUSTED_DOMAIN_INFORMATION_EX domainI
 
         internal static string CreateTrustPassword()
         {
-            string password = string.Empty;
-            byte[] buf = new byte[PASSWORD_LENGTH];
+#if NETCOREAPP3_0_OR_GREATER
+            return string.Create<object?>(PASSWORD_LENGTH, null, static (destination, _) =>
+            {
+                for (int i = 0; i < destination.Length; i++)
+                {
+                    destination[i] = PasswordCharacterSet[RandomNumberGenerator.GetInt32(PasswordCharacterSet.Length)];
+                }
+            });
+#else
             char[] cBuf = new char[PASSWORD_LENGTH];
 
-            using (RandomNumberGenerator RNG = RandomNumberGenerator.Create())
+            using (RandomNumberGenerator rng = RandomNumberGenerator.Create())
             {
-                RNG.GetBytes(buf);
-                for (int iter = 0; iter < PASSWORD_LENGTH; iter++)
+                for (int i = 0; i < PASSWORD_LENGTH; i++)
                 {
-                    int i = (int)(buf[iter] % 87);
-                    if (i < 10)
-                        cBuf[iter] = (char)('0' + i);
-                    else if (i < 36)
-                        cBuf[iter] = (char)('A' + i - 10);
-                    else if (i < 62)
-                        cBuf[iter] = (char)('a' + i - 36);
-                    else
-                        cBuf[iter] = s_punctuations[i - 62];
+                    cBuf[i] = PasswordCharacterSet[GetRandomByte(rng, PasswordCharacterSet.Length)];
                 }
             }
 
-            password = new string(cBuf);
+            return new string(cBuf);
+
+            // This is a private copy of RandomNumberGenerator.GetInt32 that works
+            // on platforms that don't have it. Since this is for a private specific
+            // use, it can be simplified to assume some things about the ranges.
+            // We know the upper-bound is < 255, so we can avoid converting random
+            // bytes to integers and mask a single byte instead. We also know the
+            // lower-bound is 0.
+            static int GetRandomByte(RandomNumberGenerator rng, int toExclusive)
+            {
+                Debug.Assert(toExclusive > 0 && toExclusive < byte.MaxValue);
+
+                // The total possible range is [0, 255).
+                int range = toExclusive - 1;
 
-            return password;
+                // Create a mask for only the applicable bits.
+                int mask = range;
+                mask |= mask >> 1;
+                mask |= mask >> 2;
+                mask |= mask >> 4;
+                // Don't need >> 8 or >> 16 since it is known the range is less than 255.
+
+                byte[] randomByte = new byte[1];
```

Comment:
this is now allocating a new array per byte?

Reply:
Hm, yeah. It's only for framework targets that don't have Span friendly random APIs, but on the other hand it's easy to just pull it out of the local method, so, fixed.